util/rateLimit.go

```diff
@@ -0,0 +1,470 @@
+	serves           []time.Time
+}
+
+// NewREDCongestionManager creates a Congestion Manager which will watches capacityGuard activity,
+// and regularly calculates a Target Service Rate, and can give "Should Drop" suggestions
+func NewREDCongestionManager(d time.Duration, r int) *redCongestionManager {
+	ret := redCongestionManager{
+		runLock:                &deadlock.Mutex{},
+		window:                 d,
+		consumed:               make(chan event, 100000),
+		served:                 make(chan event, 100000),
+		shouldDropQueries:      make(chan shouldDropQuery, 100000),
+		targetRateRefreshTicks: r,
+		consumedByClient:       map[ErlClient]*[]time.Time{},
+		exp:                    4,
+	}
+	return &ret
+}
+
+// Consumed implements CongestionManager by putting an event on the consumed channel,
+// to be processed by the Start() loop
+func (cm redCongestionManager) Consumed(c ErlClient, t time.Time) {
+	select {
+	case cm.consumed <- event{
+		c: c,
+		t: t,
+	}:
+	default:
+	}
+}
+
+// Served implements CongestionManager by putting an event on the done channel,
+// to be processed by the Start() loop
+func (cm redCongestionManager) Served(t time.Time) {
+	select {
+	case cm.served <- event{
+		t: t,
+	}:
+	default:
+	}
+}
+
+// ShouldDrop implements CongestionManager by putting a query shouldDropQueries channel,
+// and blocks on the response to return synchronously to the caller
+// if an error should prevent the query from running, the result is defaulted to false
+func (cm redCongestionManager) ShouldDrop(c ErlClient) bool {
+	ret := make(chan bool)
+	select {
+	case cm.shouldDropQueries <- shouldDropQuery{
+		c:   c,
+		ret: ret,
+	}:
+		return <-ret
+	default:
+		return false
+	}
+}
+
+// Start will kick off a goroutine to consume activity from the different activity channels,
+// as well as service queries about if a given capacityQueue should drop
+func (cm *redCongestionManager) Start(ctx context.Context, wg *sync.WaitGroup) {
+	// check if the maintainer is already running to ensure there is only one routine
+	cm.runLock.Lock()
+	defer cm.runLock.Unlock()
+	if cm.running {
+		return
+	}
+	cm.running = true
+	go cm.run(ctx, wg)
+}
+
+func (cm *redCongestionManager) run(ctx context.Context, wg *sync.WaitGroup) {
+	tick := 0
+	targetRate := float64(0)
+	consumedByClient := map[ErlClient]*[]time.Time{}
+	serves := []time.Time{}
+	exit := false
+	for {
+		// first process any new events happening
+		select {
+		// consumed events -- a client has consumed capacity from a queue
+		case e := <-cm.consumed:
+			if consumedByClient[e.c] == nil {
+				ts := []time.Time{}
+				consumedByClient[e.c] = &ts
+			}
+			*(consumedByClient[e.c]) = append(*(consumedByClient[e.c]), e.t)
+		// served events -- the capacity has been totally served
+		case e := <-cm.served:
+			serves = append(serves, e.t)
+		// "should drop" queries
+		case query := <-cm.shouldDropQueries:
+			cutoff := time.Now().Add(-1 * cm.window)
```

Comment:
now().sub(cm.window) ?

Reply:
time.Sub doesn't support durations, but time.Add does:

https://go.dev/play/p/iIQqRnpXjyn

If you uncomment the time.Sub line you should see what I mean

Reply:
fair